specification: CoVE-IO theory of operations

sameo opened the pull request:
Describe CoVE-IO ABI operations, including platform setup, device initialization and interface assignment.

sameo commented:
cc @rsahita

rsahita commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,282 @@
+
+The CoVE-IO specification extends the CoVE host and guest SBI extensions to
+allow TVMs to establish trust with TEE-IO devices, and then use and interact
+with those devices. Untrusted domain components are responsible for assigning
```
```suggestion
with those devices. Untrusted supervisor domain components are responsible for assigning
```

rsahita commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,282 @@
+of the assigned devices. The TSM, on the other hand, establishes and maintains
+the secure physical and logical links between TVMs and their assigned devices.
+
+The following sections describe the functionality of the TSM provided CoVE-IO
```
```suggestion
The following sections describe the functionality of the TSM-provided CoVE-IO
```

rsahita commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+`sbi_covh_notify_iommu_msi()` COVH function while there are no pending MSI from
+the trusted IOMMU or not invoking it when there actually is one such pending
+interrupt. In the former case, the TSM can verify that there really is a pending
+MSI by checking the trusted IOMMU status registers. The latter case could cause
```
have we introduced what we mean by "trusted IOMMU status registers"?

sameo replied:
We have not, I'll rephrase.

rsahita commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+the trusted IOMMU or not invoking it when there actually is one such pending
+interrupt. In the former case, the TSM can verify that there really is a pending
+MSI by checking the trusted IOMMU status registers. The latter case could cause
+a denial of service, which is not covered by the confidential computing threat
```
```suggestion
a denial of service, which is not in scope of the confidential computing threat
```

sameo replied:
Done, thanks.

jyao1 commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+participant TSM
+participant DSM as Device DSM or RoT
+
+VMM ->> TSM: [COVH] - sbi_covh_create_spdm_session()
```
It is a good start.

We may consider how to enable IDE e.g. IDE_KM protocol.

rsahita commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+trusted device for which the DSM is the platform RoT. The SPDM establishment
+thus happens between the TSM and the platform RoT. By having the RoT playing the
+RP DSM role, the RP vendor-specific IDE key management interface is abstracted
+by the RoT.
```
do we have a figure for this?

jyao1 commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,278 @@
+TSM ->> VMM: [COVH] - spdm_session_created()
+....
+
+=== Device Initialization
```
I feel SPDM session should be part of "Device Initialization".

"Platform Initialization" should handle host side only (such as IOMMU), but not device side.

rsahita reviewed (changes requested):
some comments and change suggestions

jyao1 commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,278 @@
+DSM ->> VMM: [DOE] - SPDM_VERSION
+VMM ->> TSM: [COVH] - spdm_resp(VERSION)
+
+TSM ->> TSM: Generate SPDM request: GET_CAPS
```
I recommend we use the name aligning with SPDM specification. E.g.
GET_CAPS => GET_CAPABILITIES
CAPS => CAPABILITIES
NEGOTIATE_ALGS => NEGOTIATE_ALGORITHMS
ALGS => ALGORITHMS
KEY_RESP => KEY_EXCHANGE_RSP
FINISH_RESP => FINISH_RSP

sameo replied:
Done, thanks.

jyao1 commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,278 @@
+DSM ->> VMM: [DOE] - FINISH_RESP
+VMM ->> TSM: [COVH] - spdm_resp(FINISH_RESP)
+
+Note over VMM,DSM: TSM: SPDM Session Created
```
When do you think we need to call SPDM GET_MEASUREMENTS?

In this API, or separate API?

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,282 @@
+the secure physical and logical links between TVMs and their assigned devices.
+
+The following sections describe the functionality of the TSM provided CoVE-IO
+extensions to support trusted I/O on CoVE enabled platforms.
```
Done, thanks.

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+assigned TDIs. Only the TSM can access and program he trusted domain IOMMU
+instances.
+
+Trusted domain assigned IOMMUs can generate MSIs in order to signal the TSM
```
Done, thanks.

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+manager, e.g. the host VMM. As a consequence, it is the untrusted domain
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
```
Done, thanks.

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+7. The TSM verifies that there is a pending MSI by reading the IOMMU `s_ipsr`
+   register, and handles the interrupts as needed.
+
+The untrusted driver can trick the TSM by either invoking the
```
Done, thanks.

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+1. The TSM is loaded and provisioned with a CoVE-IO manifest.
+2. The untrusted domain manager (e.g. the host VMM) enumerates all platform
+   IOMMUs as PCIe devices.
+3. The untrusted IOMMU driver is loaded and probed.
```
Agreed. Done now, thanks.

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,277 @@
+responsibility to reserve the MSI addresses and then rely on the TSM to program
+the IOMMUs with those reserved addresses. This IOMMU registration process is
+driven by the untrusted domain manager for all IOMMUs that participate in TEE-IO
+and gies as follows:
```
Done, thanks.

jyao1 commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,430 @@
+
+1. The TSM is loaded into a supervisor domain and provisioned with a CoVE-IO
+   manifest. It is recommended that the TSM is measured by the platform
+   root-of-trust for subsequent attestation.
```
Question: Why TSM is measured by platform RoT, but not silicon RoT (CPU)?

rsahita replied:
perhaps we should say SoC RoT?

jyao1 replied:
I think it depends on who is Root-of-Trust for measurement (RTM), and who loads TSM.
If TSM is loaded by host CPU, the SoC RoT makes more sense.
Unless TSM is really loaded by platform RoT, then platform RoT needs to measure it. And the platform RoT also needs to guarantee that the one CPU executing on is the same as the one loaded and measured by Platform RoT.

If we support both use case, then we can generalize it by saying: TSM is measured by Root-of-Trust for measurement (RTM). Maybe...

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,430 @@
+=== Platform Initialization
+
+==== IOMMU Registration and Setup
+
```
Add non-normative sections related to IOMMU spec

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,430 @@
+As described in the threat model and requirement sections, the following trusted
+I/O rules apply:
+
+* As PCIe switches are outside of the TVM TCB, only selective IDE streams are
```
Do we need to program the switches when using selective IDE streams?

jyao1 replied:
For example, IDE has register - "IDE Control Register (Offset 08h)"
BIT2 - **Flow-Through IDE Stream Enable** - For Switch Ports and Root Ports, Enables the Port for flow-through operation of TLPs associated with Selective IDE streams.

jyao1 commented:
General question: Do we also need consider other flow, such as interface removal? key refresh?

sameo commented on specification/theory_operations.adoc:
```diff
@@ -1,2 +1,430 @@
+
+* As PCIe switches are outside of the TVM TCB, only selective IDE streams are
+  used to protect the PCIe link between the TVM and the TDI.
+* A single selective IDE stream is established between the TSM and the physical
```
Fixed, thanks.

sameo commented on specification/coveio_abi.adoc:
```diff
@@ -1,2 +1,46 @@
+                                        unsigned long stream_id);
+-----
+
+Establish a secure SPDM session and configure a PCIe IDE link between the TSM
```
Correct, thanks for catching that one. Fixed now.

sameo commented:
> General question: Do we also need consider other flow, such as interface removal? key refresh?

Yes, good call. I had Interface unassignment and Device disconnection in mind. Others:

- IDE key referesh
- Interface removal
- Device removal

What else?

jyao1 commented:
> What else?

Such SPDM key update, SPDM heartbeat.

jyao1 commented:
I think it is good enough to merge now.

The rest of feedback can be tracked in other issues.

sameo commented:
@rsahita I think I addressed most of your comments, so I'll go ahead and merge that PR. It's incomplete, so we'll open new issues for all the flows we want to document in this section and then we'll complete it step by step.
